File: 23-merge-k-sorted-lists/merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

struct compare {
    bool operator()(const ListNode* l, const ListNode* r) {
        return l->val > r->val;
    }
};
class Solution {
public:
    ListNode *mergeKLists(vector<ListNode *> &lists) { //priority_queue
        priority_queue<ListNode *, vector<ListNode *>, compare> q;
        // push all current heads.
        for(auto l : lists) {
            if(l != nullptr)  {
                q.push(l);
            }
        }

        if(q.empty())  return NULL;

        ListNode* result = q.top();
        q.pop();


        if(result->next) q.push(result->next);

        ListNode* curr = result;

        while(!q.empty()) {
            curr->next = q.top();
            q.pop();
            curr = curr->next;
            if(curr->next) q.push(curr->next);
        }
        return result;
    }
};

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a.val));
        for (ListNode node : lists) {
            if (node != null)
                pq.offer(node);
        }
        ListNode dummy = new ListNode(-1);
        ListNode curr = dummy;

        while (!pq.isEmpty()) {
            ListNode currNode = pq.poll();
            curr.next = new ListNode(currNode.val);
            if (currNode.next != null) {
                pq.offer(currNode.next);
            }
            curr = curr.next;
        }
        return dummy.next;

    }
}